### ISSUE TITLE
How should I use the “rect” parameter in the YOLO training, validation, and prediction processes?

### ISSUE THREAD

**Issue by enescicek** [ultralytics/ultralytics#24467]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am working on medical images, and the image sizes I work with are usually quite large. The images are rectangular, and in most cases the horizontal side is longer. I have a few questions related to this topic:

1. In YOLO train mode, should I use the `rect` parameter as `rect=True` or `rect=False`? What are the differences between these two usage scenarios?

2. Why does the default value of the `rect` parameter change depending on the `train`, `val`, and `pred` modes?

3. If I use `rect=True` during training but `rect=False` during prediction, would that cause any issues? Is this a correct way to use it?

4. Is it wrong to use the `rect` parameter as all `True` or all `False` for the `train`, `val`, and `pred` modes?

5. Is there any issue with using `imgsz=1280 rect=True`, or should I provide the actual image dimensions instead? For example, should I use something like `imgsz=640,1280 rect=True`?

6. I read the documentation about what the `rect` parameter does, but I could not fully understand it. Could you please explain it in more detail here as well?

Thank you in advance for your help.
Best regards.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @enescicek, thank you for your interest in Ultralytics 🚀! This is an automated response to help get you started, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a custom training and usage ❓ question, please provide as much information as possible, including:
- Your current `train`, `val`, and `predict` command or Python code
- A few example image shapes from your dataset
- Any training logs or screenshots that show how you are currently using `rect`
- Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips)

If this becomes a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
@enescicek `rect` is one of those flags where the default makes sense for typical training and almost never needs touching, but it's worth knowing what it actually flips. Here's the short version, based on the code as it sits today.

**During training**, `rect=True` sorts your images by aspect ratio, packs each batch into a rectangle that fits everything in it (padded to a stride multiple, default 32), and turns off mosaic, mixup, and cutmix. It also forces `shuffle=False` and won't work on Multi-GPU - the trainer just warns and flips it back to `False`. On coco8 with `imgsz=640 batch=4 rect=True`, I got per-batch tensors of `(3, 480, 640)` instead of `(3, 640, 640)`, so about a 25% saving on pixel work. The catch is you're trading that for mosaic, which on most datasets is worth more mAP than what rect saves.

**During validation**, two things to know. If you run `yolo val ...` standalone, the default is `rect=False`, and non-PT/dynamic backends force it off anyway. But during training, the val dataloader is always built with `rect=True` regardless of what you pass on the CLI - that's hardcoded in `detect/train.py:77`. So you don't really need to do anything for val.

**During prediction**, rect doesn't sort or rebatch anything. It just toggles `LetterBox(auto=True)`, which pads to a stride multiple instead of a full square - and only when every image in the batch has the same shape and your model is PT or dynamic. Mixed-shape batches fall back to square letterboxing.

Now to your specific questions:

1. For medical images that are usually wide, try both. If the dataset is dominated by similar aspect ratios and you're memory-constrained, `rect=True` is genuinely useful. If you have room to spare and care about mAP, the default `rect=False` plus mosaic will probably win.

2. The defaults differ because each mode has different priorities. Training wants mosaic more than it wants padding savings. Val during training wants speed and has no augmentations to lose. Predict doesn't assume your batch is homogeneous.

3. Train with `rect=True`, predict with `rect=False`? Totally fine, they're independent. The weights don't remember how you padded during training.

4. All `True` is fine if your aspect ratios are uniform, but understand you're giving up mosaic and Multi-GPU. All `False` is the safe baseline.

5. `imgsz=640,1280` doesn't work for training or val - it gets silently coerced to `imgsz=1280` with a warning. Two-value imgsz only works for predict and export. With `rect=True imgsz=1280`, the rectangular shape is figured out per batch from your actual image aspect ratios. Just give it the long side as a single int.

6. The mental model: without rect, every image is letterboxed to a square. With rect, each batch picks its own non-square shape based on what's actually in it. Saves pixels, costs you mosaic.

**Comment by enescicek** [comment]
Why can't I use mosaic augmentation when “rect=True” is set? What is the reason for this?

**Comment by raimbekovm** [comment]
@enescicek good follow-up. The short answer is that the two modes attach contradictory shape info to each sample, and the trainer just disables mosaic before that contradiction can blow up.

When you set `rect=True`, the dataset sorts every image by aspect ratio, groups them into batches, and computes one rectangular shape per batch. Then on every fetch it stamps that shape onto the label as `rect_shape` so LetterBox knows to pad to a non-square rectangle. I checked this on coco8 with `imgsz=640 batch=4 rect=True` and got `rect_shape=[512, 672]` on the labels. With `rect=False` the same field is just `None`.

Mosaic, on the other hand, pulls 3, 4, or 9 random images from anywhere in the dataset and pastes them onto a single composite canvas. If the dataset is in rect mode, every one of those random images is carrying its own `rect_shape`, but the composite output is one roughly-square image with one shape. The mosaic op refuses to deal with that and just asserts:

```
assert labels.get("rect_shape") is None, "rect and mosaic are mutually exclusive."
```

(`data/augment.py:549`). I forced this path manually to make sure: built a Mosaic, set `rect_shape` on a label, ran `_mix_transform`, and got exactly that AssertionError.

The trainer never lets you hit that assert though, because `data/dataset.py:216-218` zeros out `hyp.mosaic`, `hyp.mixup`, and `hyp.cutmix` whenever `self.rect` is True. The Mosaic transform is still constructed inside `v8_transforms`, just with `p=0.0`, so it gets skipped every time. I verified this too: with `rect=False` the Mosaic inside the inner Compose has `p=1.0`, with `rect=True` it has `p=0.0`.

There is also a more intuitive way to see why: rect is trying to save padding by packing each batch into a tight non-square rectangle, and mosaic is doing the opposite by stitching 4 images into a square canvas. Running both would mean rect sorts the dataset, picks a rectangle, and then mosaic immediately throws that away and gives you a square back. You would pay for the sorting and get none of the savings.

So it is not really a missing feature, it is a deliberate choice that keeps the two augmentation paths from stepping on each other.

**Comment by enescicek** [comment]
Thank you for the detailed explanation.

I have one more question about this. In the current Ultralytics implementation, does the mosaic augmentation necessarily require a square canvas by design?

In theory, could mosaic be implemented with a rectangular canvas that matches the rect batch shape, or would that conflict with the current LetterBox / rect_shape logic?

I understand that `rect=True` and mosaic are currently mutually exclusive, but I am trying to understand whether this is mainly because of the current implementation choice, or because mosaic fundamentally needs a square canvas to work properly.

Thanks again for your help.

**Comment by raimbekovm** [comment]
@enescicek the current implementation is square at every step, not because of mosaic math but because the pipeline is wired that way. I checked the code and ran it on coco8 to confirm:

1. `Mosaic.__init__` takes `imgsz` as a single `int`, not an `(h, w)` pair.
2. `self.border = (-imgsz // 2, -imgsz // 2)` is symmetric in H and W.
3. `_mosaic4` allocates the canvas as `np.full((s * 2, s * 2, ...))`. `_mosaic9` allocates `(s * 3, s * 3, ...)` and crops back to `2s x 2s`. Both produce a `(1280, 1280, 3)` array on `imgsz=640`.
4. Right after mosaic, `RandomPerspective` is built with `pre_transform=LetterBox(new_shape=(imgsz, imgsz))`, so the output is letterboxed to a square `(imgsz, imgsz)` regardless.

So mosaic does not need a square canvas as a matter of geometry, the bbox math would work the same on a rectangular canvas. The square shape is fixed by these four hardcoded points and by the downstream LetterBox, which is what makes it incompatible with `rect_shape`.

**Comment by enescicek** [comment]
Thank you for your response

**Comment by glenn-jocher** [comment]
You're welcome — glad that clarified it. If any follow-up comes up while testing, please open a new [GitHub Discussion](https://github.com/orgs/ultralytics/discussions) and we’ll take a look.
